src/mbgl/style/expression/number_format.cpp

```diff
@@ -0,0 +1,293 @@
+    std::size_t length = arrayLength(value);
+
+    if (length != 3) {
+        ctx.error("Expected two arguments, but found " + util::toString(length) + " instead.");
+        return ParseResult();
+    }
+
+    ParseResult number = ctx.parse(arrayMember(value, 1), 1, {type::Number});
+    if (!number) {
+        ctx.error("Failed to parse the number.");
+        return ParseResult();
+    }
+
+    type::Type type = (*number)->getType();
+    if (!type.is<type::NumberType>()) {
+        ctx.error("Expected argument of type number, but found " + toString(type) + " instead.");
+        return ParseResult();
+    }
+
+    auto options = arrayMember(value, 2);
+    if (!isObject(options)) {
+        ctx.error("Number-format options argument must be an object.");
+        return ParseResult();
+    }
+
+    const optional<Convertible> localeOption = objectMember(options, "locale");
+    ParseResult locale;
+    if (localeOption) {
+        locale = ctx.parse(*localeOption, 1, {type::String});
+        if (!locale) {
+            ctx.error("Number-format locale parsing failed.");
+            return ParseResult();
+        }
+    }
+
+    const optional<Convertible> currencyOption = objectMember(options, "currency");
+    ParseResult currency;
+    if (currencyOption) {
+        currency = ctx.parse(*currencyOption, 1, {type::String});
+        if (!currency) {
+            ctx.error("Number-format currency parsing failed.");
+            return ParseResult();
+        }
+    }
+
+    const optional<Convertible> minFractionDigitsOption = objectMember(options, "min-fraction-digits");
+    ParseResult minFractionDigits;
+    if (minFractionDigitsOption) {
+        minFractionDigits = ctx.parse(*minFractionDigitsOption, 1, {type::Number});
+        if (!minFractionDigits) {
+            ctx.error("Number-format min-fraction-digits parsing failed.");
+            return ParseResult();
+        }
+    }
+
+    const optional<Convertible> maxFractionDigitsOption = objectMember(options, "max-fraction-digits");
+    ParseResult maxFractionDigits;
+    if (maxFractionDigitsOption) {
+        maxFractionDigits = ctx.parse(*maxFractionDigitsOption, 1, {type::Number});
+        if (!maxFractionDigits) {
+            ctx.error("Number-format max-fraction-digits parsing failed.");
+            return ParseResult();
+        }
+    }
+
+    return ParseResult(std::make_unique<NumberFormat>(std::move(*number),
+                                                      std::move(locale),
+                                                      std::move(currency),
+                                                      std::move(minFractionDigits),
+                                                      std::move(maxFractionDigits)));
+}
+
+mbgl::Value NumberFormat::serialize() const {
+    std::vector<mbgl::Value> serialized{{ getOperator() }};
+    serialized.emplace_back(number->serialize());
+
+    std::unordered_map<std::string, mbgl::Value> options;
+    if (locale) {
```

Comment:
No need to check both, because it is taken care by the optional template:

  explicit constexpr operator bool() const noexcept {
    return ref != nullptr;
  }

Reply:
but you've a unique_ptr inside optional, the `unique_ptr` is itself nullable